Revised documentation (#161)
Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard: README.md

# Green Algorithms HPC Dashboard: Deployment notes

![Version: Gamma (pre-release)](https://img.shields.io/badge/version-Gamma_(pre--release)-orange)

> The Green Algorithms HPC Dashboard helps HPC sysadmins and researchers track the energy use and carbon footprint of their cluster over time, using [Green Algorithms](https://www.green-algorithms.org). It combines a data collection backend, a PostgreSQL database, and a Grafana server that queries the database and visualises the results.

Reviewer: ```suggestion
The Green Algorithms HPC Dashboard helps users of Research Computing Infrastructure track the energy use and carbon footprint of their computing jobs over time, using the [Green Algorithms](https://www.green-algorithms.org) methodology. The only data used are the SLURM logs of individual jobs. It combines a data collection backend, a PostgreSQL database, and a Grafana server that queries the database and visualises the results.

```

By the end of this guide, you will have:

- A PostgreSQL database populated with your HPC usage data
- A running Grafana server with dashboards configured
- User accounts set up for your team

The system has three components:

> 1. Backend - queries the HPC scheduler (via sacct), aggregates usage per user per day, and calculates estimated energy use and carbon footprint

Reviewer: ```suggestion
1. Backend - queries the HPC scheduler (here `SLURM` via `sacct`), aggregates usage per user per day, and calculates estimated energy use and carbon footprint.
```

---
## Prerequisites

### 1. Python environment (Miniforge)

>  First, set up a Python environment (Python 3.13+ recommended) using [Miniforge](https://conda-forge.org/download/$0) (recommended). Follow the instructions for your platform (see also the [Miniforge GitHub repository](https://github.com/conda-forge/miniforge$0) if needed).

Reviewer 2: Typos ?
`https://conda-forge.org/download/$0`
`https://github.com/conda-forge/miniforge$0`

Author: Yes my bad. Fixed. Commit. [d9a567c](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/d9a567c0cd2c1a9959414f6c523e236cd10d3d15)

    ```
    $ python -m pip install -e .
    ```
    This option is only needed if you intend to modify the package.

### 3. PostgreSQL Database server
Install PostgreSQL locally or have access to a PostgreSQL server.

It is assumed that the operating system used to run the dashboard will be a flavour of UNIX/Linux. However, if you want to run it on a Mac, we suggest you use the relevant [Enterprise DB installer](https://www.enterprisedb.com/downloads/postgres-postgresql-downloads) to start with. Regardless, follow the instructions for your system.

>[!NOTE]
> These instructions assume PostgreSQL is installed on the same machine as the dashboard and Grafana. If your setup requires PostgreSQL to be hosted on a separate server, you may need a version of PostgreSQL compiled with SSH support.

Choose a username and password for the PostgreSQL admin user (conventionally `postgres`). These instructions assume you use `postgres` as the admin username.

> [!IMPORTANT] 
> Do not store the admin password in a file. If you forget it, follow [these steps](https://stackoverflow.com/questions/14588212/postgresql-resetting-password-of-postgresql-on-ubuntu) to reset it.

> [!NOTE]

> > Grafana users should be granted only SELECT permission on the database.

Reviewer: Do you mean dashboard users?

Author: Yes. Using 'Grafana' here specifically talk about users of the Grafana app that need dashboard access.

Reviewer: Let's stick to always calling them the same thing for consistency, "Dashboard users" is good I think

Author: Done. [fc2340b](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/fc2340b6e178480427ee282eb0332cd734149354)

Check that your `$PATH` allows you to access the `psql` utility.

### 4. Grafana

Install the [self-managed Grafana Enterprise edition](https://grafana.com/grafana/download?pg=get&plcmt=selfmanaged-box1-cta1) (Enterprise, just in case you want to host it on the cloud).

By default, the superuser credential is id: `admin`, password: `admin`. You should change the password to make your set-up more secure.

---
## Configuration files

Before running the dashboard, you will need to prepare four configuration files. Templates and examples for all of them are provided in the `configuration/templates/` and `configuration/examples/` directories respectively.

### 1. Scripts configuration file (`config.yaml`)
Copy the template provided in `configuration/templates/config.yaml` to a location of your choice and edit it, replacing all values surrounded by `< >` characters. This file contains the core parameters needed to run the dashboard, including database connection details and paths to the other configuration files. You can uncomment the optional parameters you want to use.

> [!NOTE]

> > If you choose to run the dashboard inside a Docker container, note that restarting the container can cause the database to be deleted and re-created. To prevent this, set `skip_db_overwrite: True` in `config.yaml`, but only after the first run, as the initial installation requires the database to be created.

Reviewer: The first time, the database shouldn't exist, so `skip_db_overwrite: True` should still be ok... or is it not how it works?

Author: Yes if the db does not exist, the value of skip_db_overwrite does not matter, it can be kept as True. Will change this.

Author: Done. commit [6a79cb0](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/6a79cb09999329d1f89373e695fe184818581067)

#### Carbon intensity (CI)
The dashboard uses carbon intensity (CI) data to estimate the carbon footprint of your HPC usage. There are three ways to configure this:

1. **Carbon Intensity API (UK only)**
If your cluster is based in the UK, the dashboard can pull dynamic carbon intensity data from the [Carbon Intensity API](https://carbonintensity.org.uk). To enable this, add the first three characters of your **postcode** to `config.yaml`. The dashboard uses a daily average of the carbon intensity values returned by the API.

> 2. **Static carbon intensity value**

Reviewer: ```suggestion
2. **Fixed carbon intensity value**
```

> For clusters outside the UK, you can provide a fixed carbon intensity value directly in `config.yaml`. This will be used for all calculations.

Reviewer: ```suggestion
For clusters outside the UK, you can provide a fixed carbon intensity value directly in `config.yaml`, e.g. the average carbon intensity in the data centre's location (you can find it [here](https://app.electricitymaps.com)). This will be used for all calculations.
```

3. **Custom API integration**
If you would like to use a different carbon intensity API for your region, you can implement your own integration using the `APIService` class in `ga_dashboard/backend/services/api_service.py`. Get in touch if you need guidance on this.

### 2. Cluster information file (`cluster_info.yaml`)
Copy the template provided in `configuration/templates/cluster_info.yaml` and edit it with information about your HPC cluster. See `configuration/examples/cluster_info__demo.yaml` for a worked example.

For each partition (a set of computing nodes with a dedicated queue) you will need:
- `type`: CPU or GPU
- `model`: processor model
- `TDP`: thermal design power (check the manufacturer's datasheet if needed)
- For GPU partitions only: `model_CPU` and `TDP_CPU`

You will also need values for `institution`, `cluster_name`,`granularity_memory_request`, `PUE`, and other fields listed in the template.

### 3. Fixed parameters file (`fixed_parameters.yaml`) 

> Use the provided file at `ga_dashboard/data/fixed_parameters.yaml` without modification for now.

Reviewer: If they don't have to touch it, why mention it here?

Author: Right, kept it here incase the user wants to change some fixed parameter for any reason - I mean if they have a value they think is better suited. Can leave it out for now I think.

Author: Removed. commit [6a79cb0](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/6a79cb09999329d1f89373e695fe184818581067)

### 4. Dashboard users file (`user_list.csv`)
This file lists the users who will have access to the dashboard. Copy the template provided in `configuration/templates/user_list.csv` and populate with your users' details.

The users file should be a comma-separated file combining these columns:
* **User name**: Company/Institute user name (e.g. tg1)

> * **User unique identifier (UID)**: Numeric user unique identifier (e.g. 11111)

Reviewer: We need to make clear how this is being mapped to SLURM logs

Author: Added an important note for this. I confirmed from the code that the mapping is being done on User. 

`df2 = pd.merge(df, self.users_df, left_on='UserX', right_on='User', how='inner')`
[go to code](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/blob/88c1160018b7a82186076e5228e703634bb0e926/ga_dashboard/backend/ga_tools.py#L253)

commit [6a79cb0](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/6a79cb09999329d1f89373e695fe184818581067)

#### Carbon intensity (CI)
The dashboard uses carbon intensity (CI) data to estimate the carbon footprint of your HPC usage. There are three ways to configure this:

1. **Carbon Intensity API (UK only)**
If your cluster is based in the UK, the dashboard can pull dynamic carbon intensity data from the [Carbon Intensity API](https://carbonintensity.org.uk). To enable this, add the first three characters of your **postcode** to `config.yaml`. The dashboard uses a daily average of the carbon intensity values returned by the API.
2. **Static carbon intensity value**
For clusters outside the UK, you can provide a fixed carbon intensity value directly in `config.yaml`. This will be used for all calculations.
3. **Custom API integration**
If you would like to use a different carbon intensity API for your region, you can implement your own integration using the `APIService` class in `ga_dashboard/backend/services/api_service.py`. Get in touch if you need guidance on this.

> ### 2. Cluster information file (`cluster_info.yaml')

Reviewer 2: Typo: second quote character is wrong:
> Cluster information file (`cluster_info.yaml')

Author: Fixed. Commit. [d9a567c](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/d9a567c0cd2c1a9959414f6c523e236cd10d3d15)

Copy the template provided in `configuration/templates/cluster_info.yaml` and edit it with information about your HPC cluster. See `configuration/examples/cluster_info__demo.yaml` for a worked example.

For each partition (a set of computing nodes with a dedicated queue) you will need:
- `type`: CPU or GPU
- `model`: processor model
- `TDP`: thermal design power (check the manufacturer's datasheet if needed)
- For GPU partitions only: `model_CPU` and `TDP_CPU`

You will also need values for `institution`, `cluster_name`,`granularity_memory_request`, `PUE`, and other fields listed in the template.

### 3. Fixed parameters file (`fixed_parameters.yaml`) 
Use the provided file at `ga_dashboard/data/fixed_parameters.yaml` without modification for now.

### 4. Dashboard users file (`user_list.csv`)
This file lists the users who will have access to the dashboard. Copy the template provided in `configuration/templates/user_list.csv` and populate with your users' details.

The users file should be a comma-separated file combining these columns:
* **User name**: Company/Institute user name (e.g. tg1)

> * **User unique identifier (UID)** (UID): Numeric user unique identifier (e.g. 11111)

Reviewer 2: `UID` written twice

Author: Fixed. Commit. [d9a567c](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/d9a567c0cd2c1a9959414f6c523e236cd10d3d15)

| ...   | ...   | ...           | ...     | ...  | ...      | ... |

> [!IMPORTANT]
> Passwords must not contain a comma character (`','`), as this will break CSV parsing.

> ##

Reviewer: Can be deleted?

Author: Done. commit [6a79cb0](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/6a79cb09999329d1f89373e695fe184818581067)

---

> ## Install Green Algorithms dashboard

Reviewer: ```suggestion
## Install the Green Algorithms dashboard
```

| ...   | ...   | ...           | ...     | ...  | ...      | ... |

> [!IMPORTANT]
> Passwords must not contain a comma character (`','`), as this will break CSV parsing. Passwords should also adhere to the [Grafana password policy](https://grafana.com/docs/grafana/next/setup-grafana/configure-security/configure-authentication/grafana/#strong-password-policy), should you decide to enforce it.

> ##

Reviewer 2: Could be removed ?

Author: I've removed the second part of the sentence - kept the 'csv will break' bit.

Commit. [d9a567c](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/d9a567c0cd2c1a9959414f6c523e236cd10d3d15)

Reviewer 2: I think it was about the line `##`

```
This will prompt you to enter the PostgreSQL user password and then the Grafana admin password.

If the script completes without errors, the installation was successful. If you re-run the script, you may see `WARNING` messages about existing resources (e.g. teams or users already in Grafana), these are expected and can be ignored.

---
## HPC usage data collection

This section covers how to populate the database with HPC usage data. There are two scripts depending on whether you are running the data collection for the first time or on a recurring schedule.

### First run (historical data)
To collect and process all available historical logs, run:
```
$ python scripts/run_green_algorithms_on_historical_logs.py --config <your_config_file.yaml>
```
This will collect all the logs available by default. You can restrict the date range using `startDay` and `endDay` in `config.yaml`. The logs are pulled and processed in batches (use `--batch_size` to configure number of days per batch).

### Scheduled runs

> For a scheduled run (e.g. a `cron` job) for data collection and processing, run:

Reviewer: We haven't said anywhere that the expected way for the dashboard to run is to process data daily. This should be added in the introduction at the top.

Author: Added a line at the very top. commit [6a79cb0](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/6a79cb09999329d1f89373e695fe184818581067)

```
$ python scripts/run_green_algorithms_on_logs.py --config <your_config_file.yaml>
```

> By default, it will collect the logs from previous day. The date range can be controlled using `startDay` and `endDay` in `config.yaml`.

Reviewer: We should have default values for this no? e.g. "yesterday"

Author: Yes we have that. by default its the previous day.

Author: Slightly changed the wording. commit [6a79cb0](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/pull/161/commits/6a79cb09999329d1f89373e695fe184818581067)

If you log in as an administrator, there are many other functions (e.g. delete users, add dashboards, add data sources). Full details may be made available later, perhaps as a tutorial.

---
## Getting help
If you have questions, run into issues, or want to share feedback, please open a thread in [GitHub Discussions](https://github.com/Cambridge-Sustainable-Computing-Lab/Green-Algorithms-HPCdashboard/discussions). This is the best place to get support from the development team and the wider community.

> 

Reviewer: ```suggestion

---
## About us

This tool is built and maintained by the [Cambridge Sustainable Computing Lab](https://cam-sustainablecomputing.org) at the University of Cambridge, UK. 

```
